File: app/src/main/java/android/vn/leo/qrscan/fragments/ScanFragment.java
package android.vn.leo.qrscan.fragments;

import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.vn.leo.qrscan.R;
import android.vn.leo.qrscan.interfaces.OnResult;
import android.widget.Toast;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.zxing.ResultPoint;
import com.journeyapps.barcodescanner.BarcodeCallback;
import com.journeyapps.barcodescanner.BarcodeResult;
import com.journeyapps.barcodescanner.DecoratedBarcodeView;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

public class ScanFragment extends Fragment implements DecoratedBarcodeView.TorchListener,
        BarcodeCallback, View.OnKeyListener {

    private DecoratedBarcodeView mDecoratedBarcodeView;
    private boolean isFlashOn = false;

    private OnResult resultCallback;

    private AdView mAdView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_scan, container, false);

        mDecoratedBarcodeView = view.findViewById(R.id.zxing_barcode_scanner);

        mDecoratedBarcodeView.decodeContinuous(this);
        mDecoratedBarcodeView.setTorchListener(this);
        mDecoratedBarcodeView.setOnKeyListener(this);

//        mAdView = view.findViewById(R.id.adView);
//        AdRequest adRequest = new AdRequest.Builder().addTestDevice("989E8904E0C066595F894A9EE90E0911").build();
//        mAdView.loadAd(adRequest);

        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        this.resultCallback = (OnResult) getActivity();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        MenuInflater menuInflater = getActivity().getMenuInflater();
        menuInflater.inflate(R.menu.main_menu, menu);
        if (isFlashOn) {
            MenuItem item = menu.findItem(R.id.btn_flash);
            if (item != null) {
                item.setIcon(R.drawable.ic_flash_on_24dp);
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.btn_flash: {
                onFlash();
                if (isFlashOn) {
                    item.setIcon(R.drawable.ic_flash_on_24dp);
                } else {
                    item.setIcon(R.drawable.ic_flash_off_24dp);
                }
                return true;
            }
        }
        return false;
    }

    public void onFlash() {
        isFlashOn = !isFlashOn;
        turnOnOrOffLight();
    }

    public void turnOnOrOffLight() {
        if (isFlashOn) {
            mDecoratedBarcodeView.setTorchOn();
        } else {
            mDecoratedBarcodeView.setTorchOff();
        }
    }

    @Override
    public void barcodeResult(BarcodeResult result) {
        if (this.resultCallback == null) {
            return;
        }
        if (this.resultCallback.isOnResult()) {
            return;
        }
        this.resultCallback.onResult(result);
    }

    @Override
    public void possibleResultPoints(List<ResultPoint> resultPoints) {

    }

    @Override
    public void onResume() {
        if (mDecoratedBarcodeView != null &&
                this.resultCallback != null && !this.resultCallback.isDisabled()) {
            mDecoratedBarcodeView.pause();
            mDecoratedBarcodeView.resume();
        }
        super.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        if (mDecoratedBarcodeView != null) {
            mDecoratedBarcodeView.pause();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mDecoratedBarcodeView != null) {
            mDecoratedBarcodeView.decodeContinuous(null);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    @Override
    public boolean onKey(View v, int keyCode, KeyEvent event) {
        return event.getAction() == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_BACK;
    }

    @Override
    public void onTorchOn() {
        Toast.makeText(getContext(), getResources().getString(R.string.light_flash_on), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onTorchOff() {
        Toast.makeText(getContext(), getResources().getString(R.string.light_flash_off), Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/android/vn/leo/qrscan/ui/scan/ScanFragment.kt
package android.vn.leo.qrscan.ui.scan

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.vn.leo.qrscan.R
import android.vn.leo.qrscan.base.ui.BaseFragment
import android.vn.leo.qrscan.databinding.FragmentScanBinding
import android.vn.leo.qrscan.dialog.bottomsheet.coderesult.ResultDialogDelegate
import android.vn.leo.qrscan.dialog.listener.OnDialogDismissListener
import android.vn.leo.qrscan.extensions.screenSize
import android.vn.leo.qrscan.extensions.showToast
import android.vn.leo.qrscan.extensions.toPx
import android.vn.leo.qrscan.model.BarcodeParsedResult
import android.vn.leo.qrscan.parser.BarcodeResultParser
import androidx.core.view.isVisible
import androidx.lifecycle.ViewModelProvider
import com.google.zxing.ResultPoint
import com.journeyapps.barcodescanner.BarcodeCallback
import com.journeyapps.barcodescanner.BarcodeResult
import com.journeyapps.barcodescanner.DecoratedBarcodeView
import com.journeyapps.barcodescanner.Size

class ScanFragment : BaseFragment<FragmentScanBinding>(), BarcodeCallback,
    DecoratedBarcodeView.TorchListener, OnDialogDismissListener {

    override fun onDismiss() {
        viewBinding.zxingBarcodeScanner.resume()
    }

    override fun createViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): FragmentScanBinding {
        return FragmentScanBinding.inflate(inflater, container, false)
    }

    private val scanViewModelFactory = ScanViewModelFactory(BarcodeResultParser())

    private val scanViewModel: ScanViewModel by lazy {
        ViewModelProvider(this, scanViewModelFactory)[ScanViewModel::class.java]
    }

    private val resultDialogDelegate: ResultDialogDelegate by lazy {
        ResultDialogDelegate(childFragmentManager)
    }

    override fun onRenderView(view: View, savedInstanceState: Bundle?) {
        viewBinding.zxingBarcodeScanner.barcodeView.framingRectSize =
            Size(screenSize().width - 64.toPx(), 400.toPx())
        viewBinding.zxingBarcodeScanner.decodeContinuous(this)
        viewBinding.zxingBarcodeScanner.setTorchListener(this)
        scanViewModel.handleResultEvent.observe(viewLifecycleOwner, ::handleBarcodeParsedResult)
        scanViewModel.loadingEvent.observe(viewLifecycleOwner, ::handleLoading)
        scanViewModel.toastEvent.observe(viewLifecycleOwner, ::showToast)
    }

    private fun handleBarcodeParsedResult(barcodeParsedResult: BarcodeParsedResult) {
        resultDialogDelegate.showResultDialog(barcodeParsedResult)
    }

    private fun handleLoading(showLoading: Boolean) {
        viewBinding.progressCircular.isVisible = showLoading
    }

    override fun barcodeResult(result: BarcodeResult?) {
        viewBinding.zxingBarcodeScanner.pauseAndWait()
        scanViewModel.handleBarcodeResult(result)
    }

    override fun possibleResultPoints(resultPoints: MutableList<ResultPoint>?) {

    }

    override fun onTorchOn() {
        showToast(R.string.light_flash_on)
    }

    override fun onTorchOff() {
        showToast(R.string.light_flash_on)
    }

    override fun onPause() {
        super.onPause()
        scanViewModel.validateBarcodeViewOnPause {
            viewBinding.zxingBarcodeScanner.pauseAndWait()
        }
    }

    override fun onResume() {
        super.onResume()
        scanViewModel.validateBarcodeViewOnResume {
            viewBinding.zxingBarcodeScanner.resume()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        viewBinding.zxingBarcodeScanner.decodeContinuous(null)
        viewBinding.zxingBarcodeScanner.setTorchListener(null)
    }
}